File: app/src/main/java/com/example/myapplication/activities/ReadActivity.java
package com.example.myapplication.activities;

import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.MediaController;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.VideoView;

import com.example.myapplication.R;
import com.example.myapplication.data.control.DBController;
import com.example.myapplication.data.entity.Media;
import com.example.myapplication.data.interfaces.IMediable;

import java.io.InputStream;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class ReadActivity extends AppCompatActivity {

    private DBController c;
    @BindView(R.id.buttonupd) Button button;
    @BindView(R.id.buttonplay) Button button2;
    @BindView(R.id.buttonsearch) Button button3;
    @BindView(R.id.scrollView2) ScrollView scrollView;
    @BindView(R.id.imageView2) ImageView imageView;
    @BindView(R.id.videoView) VideoView videoView;
    String s = "";
    protected Integer idInt;
    Media media;

    void Init(){
        c = new DBController(this);
        idInt = getIntent().getIntExtra("id",1);
        media = c.getMedia(idInt);
    }

    private <T extends IMediable> void readInf(T media, int id){
        media = c.getMedia1(media, id);
        s = media.getName();
    }

    protected void onCreate(Bundle savedInstanceState) {
        SharedPreferences prefs= PreferenceManager.getDefaultSharedPreferences(this);
        switch (prefs.getString("colors","Зеленый")){
            case "Серый":{setTheme(R.style.AppTheme);break;}
            case "Красный":{setTheme(R.style.Red);break;}
            case "Зеленый":{setTheme(R.style.Green);break;}
            case "Синий":{setTheme(R.style.Blue);break;}
            case "Желтый":{setTheme(R.style.Yellow);break;}
        }
        super.onCreate(savedInstanceState);
        setContentView(R.layout.music_present);
        ButterKnife.bind(this);
        imageView.setVisibility(View.INVISIBLE);
        videoView.setVisibility(View.INVISIBLE);
    }

    private class DownloadImageTask extends AsyncTask<String, Void, Bitmap> {
        ImageView bmImage;

        DownloadImageTask(ImageView bmImage) {
            this.bmImage = bmImage;
        }

        protected Bitmap doInBackground(String... urls) {
            String urldisplay = urls[0];
            Bitmap mIcon11 = null;
            try {
                InputStream in = new java.net.URL(urldisplay).openStream();
                mIcon11 = BitmapFactory.decodeStream(in);
            } catch (Exception e) {
            }
            return mIcon11;
        }

        protected void onPostExecute(Bitmap result) {
            bmImage.setVisibility(View.VISIBLE);
            bmImage.setImageBitmap(result);
        }
    }

    private class DownloadVideoTask extends AsyncTask<String, Void, Void>{
        VideoView videoView;

        DownloadVideoTask(VideoView bmImage) {
            this.videoView = bmImage;
        }

        @Override
        protected Void doInBackground(String... strings) {
            videoView.setVideoURI(Uri.parse(strings[0]));
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {

            videoView.requestFocus(0);
            videoView.setVisibility(View.VISIBLE);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        scrollView.removeAllViews();
        videoView.setVisibility(View.INVISIBLE);
    }

    @Override
    protected void onResume() {
        super.onResume();
        Init();
        setTitle(media.getName());
        readInf(media.mediaType(),media.getId());
        switch (media.getType()){
            case "Music":{
                TextView textView2 = new TextView(this);
                textView2.setText(s);
                textView2.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
                textView2.setBackgroundResource(R.drawable.side_nav_bar);

                scrollView.addView(textView2);
                break;
            }
            case "Excerption":{
                TextView textView2 = new TextView(this);
                textView2.setText(s);
                textView2.setBackgroundResource(R.drawable.side_nav_bar);
                scrollView.addView(textView2);
                break;
            }
            case "Image":{
                new DownloadImageTask(imageView)
                        .execute(media.getInsideUri());
                break;
            }
            case "Film":{
                if(media.getInsideUri()!=null && !media.getInsideUri().equals(""))
                    new DownloadVideoTask(videoView)
                            .execute(media.getInsideUri());
                videoView.setMediaController(new MediaController(this));
                //scrollView.addView(videoView);
                break;
            }
            case "Anime":{
                if(media.getInsideUri()!=null && !media.getInsideUri().equals(""))
                    new DownloadVideoTask(videoView)
                            .execute(media.getInsideUri());
                videoView.setMediaController(new MediaController(this));
                //scrollView.addView(videoView);
                break;
            }
        }

    }

    @OnClick(R.id.buttonsearch)
    void brouse(){
        Intent ii = new Intent(this, BrowseActivity.class);
        ii.putExtra("search", "yandsearch?text=" + c.getAlbum(media.getId()) + " " + media.getName());
        startActivity(ii);
    }

    @OnClick(R.id.buttonplay)
    void OnPlayClick(){
        String type ;

        switch (media.getType()){
            case "Music":{ type = "audio/*" ; break;}
            case "Image":{ type = "image/*" ; break;}
            case "Film":{ type = "video/*" ; break;}
            case "Excerption":{ type = "text/*" ; break;}
            default:{ type = "image/*";}
        }

        if ((media.getInsideUri() == null||media.getInsideUri().equals("") )&&
                ( media.getOutsideUri().equals("")||media.getOutsideUri() == null)) {
            Toast.makeText(this,"Вы не сохранили источник",Toast.LENGTH_SHORT).show();
        }
        else{
            if (media.getInsideUri()!=null && !media.getInsideUri().equals("")){
                Intent playAudioIntent = new Intent(Intent.ACTION_VIEW);
                playAudioIntent.setDataAndType(Uri.parse(media.getInsideUri()), type);
                startActivity(playAudioIntent);
            }
            else{
                Intent playAudioIntent = new Intent(Intent.ACTION_VIEW);
                playAudioIntent.setDataAndType(Uri.parse(media.getOutsideUri()), type);
                startActivity(playAudioIntent);
            }
        }
    }

    @OnClick(R.id.buttonupd)
    void OnClick(){
        Intent intent = new Intent(ReadActivity.this, AddActivity.class);
        intent.putExtra("id",media.getId());
        startActivity(intent);
    }
}

File: app/src/main/java/com/example/myapplication/activities/ReadActivity.kt
package com.example.myapplication.activities

//import com.example.myapplication.dsl.ExampleActivity

import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.preference.PreferenceManager
import android.provider.MediaStore
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.example.myapplication.MyApp.Companion.c
import com.example.myapplication.R
import com.example.myapplication.data.entity.Media
import com.example.myapplication.data.interfaces.IMediable
import kotlinx.android.synthetic.main.music_present.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.jetbrains.anko.share
import org.jetbrains.anko.startActivity
import java.net.URL
import java.net.URLEncoder

class ReadActivity : AppCompatActivity() {

    private var bitmap: Bitmap? = null
    private var s = ""
    internal lateinit var media: Media

    override fun onCreate(savedInstanceState: Bundle?) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(this)
        when (prefs.getString("colors", "Зеленый")) {
            "Серый" -> setTheme(R.style.AppTheme)
            "Красный" -> setTheme(R.style.Red)
            "Зеленый" -> setTheme(R.style.Green)
            "Синий" -> setTheme(R.style.Blue)
            "Желтый" -> setTheme(R.style.Yellow)
            "Пурпурный" -> setTheme(R.style.Purple)
        }
        super.onCreate(savedInstanceState)
        setContentView(R.layout.music_present)

        //val a = ExampleActivity().setContentView(this)
        media = c.getMedia(intent.getIntExtra("id", 1))
        imageView2.visibility = View.INVISIBLE
        videoView.visibility = View.INVISIBLE
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_readmenu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            R.id.action_search -> {

                startActivity<BrowseActivity>("search" to "yandsearch?text="+ URLEncoder.encode("${c.getAlbum(media.album)} ${media.name}", "UTF-8"))
            }
            R.id.action_update -> {
                startActivity<AddActivity>("id" to media.id)
            }
            R.id.action_share -> {
                if (bitmap == null) {
                    share(media.outsideUri)
                }else {
                    val path = MediaStore.Images.Media.insertImage(contentResolver, bitmap, "SomeText", null)
                    Log.d("Path", path)
                    val intent = Intent(Intent.ACTION_SEND)
                    val screenshotUri = Uri.parse(path)
                    intent.putExtra(Intent.EXTRA_STREAM, screenshotUri)
                    intent.type = "image/*"
                    startActivity(Intent.createChooser(intent, "Share image via..."))
                }
            }
            else -> {
                onPlayClick()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private class DownloadImageTask internal constructor(bmImage: ImageView) : AsyncTask<String, Void, Bitmap>() {
        internal var bmImage: Any = bmImage

        override fun doInBackground(vararg urls: String): Bitmap? {
            val urldisplay = urls[0]
            var mIcon11: Bitmap? = null
            try {
                val `in` = java.net.URL(urldisplay).openStream()
                mIcon11 = BitmapFactory.decodeStream(`in`)
            } catch (e: Exception) {
                Log.d("DownloadImage", e.message)
            }

            return mIcon11
        }

        override fun onPostExecute(result: Bitmap) {
            (bmImage as ImageView).visibility = View.VISIBLE
            (bmImage as ImageView).setImageBitmap(result)
        }
    }

    private class DownloadVideoTask internal constructor(bmImage: VideoView) : AsyncTask<String, Void, Void>() {
        internal var videoView: Any = bmImage

        override fun doInBackground(vararg strings: String): Void? {
            (videoView as VideoView).setVideoURI(Uri.parse(strings[0]))
            return null
        }

        override fun onPostExecute(aVoid: Void) {

            (videoView as VideoView).requestFocus(0)
            (videoView as VideoView).visibility = View.VISIBLE
        }
    }

    override fun onPause() {
        super.onPause()
        scrollView2.removeAllViews()
        videoView.visibility = View.INVISIBLE
    }

    override fun onResume() {
        super.onResume()
        title = media.name
        s = c.getMedia1(media.mediaType<IMediable>(), media.id).name
        when (media.type) {
            "Music" -> {
                val textView2 = TextView(this)
                textView2.text = s
                textView2.textAlignment = View.TEXT_ALIGNMENT_CENTER
                scrollView2.addView(textView2)
            }
            "Excerption" -> {
                val textView2 = TextView(this)
                textView2.text = s
                scrollView2.addView(textView2)
            }
            "Image" -> {
                //DownloadImageTask(imageView2).execute(media.insideUri)
                GlobalScope.launch {
                    val b = withContext(Dispatchers.Default) {
                        BitmapFactory.decodeStream(URL(media.outsideUri)
                                .openStream())
                    }
                    runOnUiThread{
                        imageView2.visibility = View.VISIBLE
                        imageView2.setImageBitmap(b)
                    }
                    bitmap = b
                }
            }
            "Film" -> {
                if (media.insideUri != null && media.insideUri != "")
                    DownloadVideoTask(videoView)
                            .execute(media.insideUri)
                videoView.setMediaController(MediaController(this))
            }//scrollView.addView(videoView);
            "Anime" -> {
                if (media.insideUri != null && media.insideUri != "")
                    DownloadVideoTask(videoView)
                            .execute(media.insideUri)
                videoView.setMediaController(MediaController(this))
            }//scrollView.addView(videoView);
        }

    }

    private fun onPlayClick() {

        val type: String = when (media.type) {
            "Music" -> "audio/*"
            "Image" -> "image/*"
            "Film" -> "video/*"
            "Excerption" -> "text/*"
            else -> "image/*"
        }

        media.insideUri?.play(type)?:media.outsideUri?.play(type)

    }

    private fun String.play(type: String):Boolean {
        if (this != "") {
            val playAudioIntent = Intent(Intent.ACTION_VIEW)
            playAudioIntent.setDataAndType(Uri.parse(this), type)
            startActivity(playAudioIntent)
            return true
        }
        Toast.makeText(this@ReadActivity, "Вы не сохранили источник", Toast.LENGTH_SHORT).show()
        return false
    }
}
